File: P006/Fortran/belief.f90
! **********************************************************
! https://justindomke.wordpress.com/2012/09/17/julia-matlab-and-c/
! **********************************************************

      program belief

      IMPLICIT NONE

      real :: r

      real start, finish

      integer N
      integer, parameter :: dim = 5000
      double precision, allocatable :: A(:,:)
      double precision, allocatable :: x(:), x2(:)
      double precision :: mysum, mynorm

      integer :: i, j, k
      CHARACTER(len=50) :: arg(1)

      ! Get the dimension from the command line
      !----------------------------------------
      call getarg(1,arg(1))
      read(arg(1), * ) N

      call cpu_time(start)

      !-------------------------------------
      ! Allocate and initialize the matrices
      !-------------------------------------
      allocate(A(dim,dim))
      allocate(x(dim))

      call srand(86456)
      do i = 1, dim
         x(i) = 1.0
         do j = 1, dim
            call random_number(r)
            A(i, j) = r
         enddo
      enddo

      !----------------------------------
      ! Perform the belief calculations
      !----------------------------------

      DO k = 1, N

         x2 = matmul(A,x)

         x(:) = log(x2(:))

         mysum = sum(exp(x(:)))

         mynorm = log(mysum)
         x(:) = x(:) - mynorm
         
      END DO

      call cpu_time(finish)

      print "('Time for Belief Calculations (',i4,'):', 2x, f15.10,'s')", N, finish - start
      !print *,'Time for Belief Calculations (',N,'): ', finish - start,' s'

      end program belief

File: P006/C/belief.c

#include <sys/types.h>
#include <time.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

const int dim = 1000;

void beliefprop2(double A[dim][dim], double x[dim], int N);

int main(int argc, char* argv[])
{
        double A[dim][dim];
        double x[dim];
        int N;
        int i, j, dim;
        clock_t start, finish;

        double maxr;

        // ---------------------------------------
        // Get the number of iterations from the command line
        // ---------------------------------------
        N = atoi(argv[1]);

        srand(time(NULL));
        start = clock();

        maxr = (double)RAND_MAX;
        for (i = 0; i < dim; i++)
        {
                x[i] = 1.0;
                for (j = 0; j < dim; j++)
                    A[i][j] = rand()/maxr;
         }

        // ---------------------------
        // Perform the belief operations
        // ---------------------------
        beliefprop2( A, x, N);
        finish = clock();

        printf("Time for belief calculations (%d): %lf s\n", N, (double) (finish - start)/CLOCKS_PER_SEC);  
        return 0;

}

void beliefprop2(double A[dim][dim], double x[dim], int N){
  int i, k, j;
  double x2[dim];
  for (k=0; k<N; k++){
      for (i=0; i<dim; i++){
          x2[i]=0;
          for (j=0; j<dim; j++)
	      x2[i] += A[i][j]*exp(x[j]);
      }   
      for (i=0; i<dim; i++)
          x[i]=log(x2[i]);
      double mysum = 0;
      for (i=0; i<dim; i++)
          mysum += exp(x[i]);
      double mynorm = log(mysum);
      for (i=0; i<dim; i++)
          x[i] -= mynorm;
  }
  return;
}

